content/posts/2017/2017-10-12-tibbles.Rmd

```diff
@@ -0,0 +1,227 @@
+```
+
+So now if there is more than 3 rows, we print only 1 row. Tibbles of length 3 and 4 would now print as so.
+ 
+ <!--  default is 20, 10-->
+
+```{r}
+tibble(1:3)
+```
+```{r}
+tibble(1:4)  
+```
+
+
+Yes, OK, you could do this with the traditional data frame. But it would be a lot more work, _right_?  
+
+As well as the fancy printing, tibbles don't drop the variable type, don't partial match and they allow non-syntactic column names when importing data in. We're going to use the data from before. Again, it is available [here](https://gist.github.com/theoroe3/8bc989b644adc24117bc66f50c292fc8). Notice it has 3 non-syntactic column names and one column of characters. Reading this is as a tibble and a data frame we get 
+
+```{r, include=FALSE}
+options(tibble.print_max = 10, tibble.print_min = 5)
+```
+
+```{r}
+tib
+```
+
+***
+
+```{r}
+df
+```
+We see already that in the `read.csv` process we've lost the column names.
+
+Let's try some partial matching...
+
+```{r}
+tib$n
+```
+```{r}
+df$n
+```
+
+With the tibble we get an error, yet with the data frame it leads us straight to our `name` variable. To read more about why partial matching is bad, check out [this thread](http://r.789695.n4.nabble.com/Deprecating-partial-matching-in-data-frame-td4661898.html).
+
+***
+
+What about subsetting? Let's try it out using the data from our csv file.
+
+```{r}
+tib[,2]
+tib[2]
+df[,2]
+df[2]
+```
+Using the a normal data frame we get a vector and a data frame using single square brackets. Using tibbles, single square brackets, `[`, will always return another tibble. Much neater.
+Now for double brackets.
+
+***
+
+```{r}
+tib[[1]]
+tib$name
+df[[1]]
+df$name
+```
+Double square brackets, `[[`, and the traditional dollar, `$` are ways to access individual columns as vectors. Now, with tibbles, we have seperate operations for data frame operations and single column operations. Now we don't have to use that pesky `drop = FALSE`. Note, these are actually quicker than the `[[` and `$` of the `data.frame`, as shown in the [documentation for the tibble package](https://cran.r-project.org/web/packages/tibble/tibble.pdf). 
+
+***
+
+At last, no more strings as factors! Upon reading the data in, tibbles recognise _strings as strings_,  not factors. For example, with the name column in our data set.
+
+```{r}
+class(df$name)
+class(tib$name)
+```
+
+I quite like this, it's much easier to turn a vector of characters into factors than vice versa, so why not give me everything as strings? Now I can choose whether or not to convert to factors. 
+
+## Disadvantages
+
+This won't be long, there's only one. Some older functions dont work with tibbles because of their different subsetting method. They expect `tib[,1]` to return a vector, when infact it will now return another data frame. Until this functionality is added in you must convert your tibble back to a data frame using `as_data_frame` or `as_tibble` as discussed previously.
```

Comment:
Link to https://stat.ethz.ch/pipermail/r-package-devel/2017q3/001896.html Also mention that this puts additional work onto package authors.